haystack_experimental/components/preprocessors/embedding_based_document_splitter.py

```diff
@@ -0,0 +1,376 @@
+
+            doc_splits = self._split_document(doc)
+            split_docs.extend(doc_splits)
+
+        return {"documents": split_docs}
+
+    def _split_document(self, doc: Document) -> List[Document]:
+        """
+        Split a single document based on embedding similarity.
+        """
+        sentences_result = self.sentence_splitter.split_sentences(doc.content)
+        sentences = [sentence["sentence"] for sentence in sentences_result]
+
+        sentence_groups = self._group_sentences(sentences)
+
+        embeddings = self._calculate_embeddings(sentence_groups)
+
+        split_points = self._find_split_points(embeddings)
+
+        splits = self._create_splits_from_points(sentence_groups, split_points)
+
+        # Merge small splits and split large splits
+        final_splits = self._post_process_splits(splits)
+
+        return self._create_documents_from_splits(final_splits, doc)
+
+    def _group_sentences(self, sentences: List[str]) -> List[str]:
+        """
+        Group sentences into groups of sentences_per_group.
+        """
+        if self.sentences_per_group == 1:
+            return sentences
+
+        groups = []
+        for i in range(0, len(sentences), self.sentences_per_group):
+            group = sentences[i : i + self.sentences_per_group]
+            groups.append(" ".join(group))
+
+        return groups
+
+    def _calculate_embeddings(self, sentence_groups: List[str]) -> List[List[float]]:
+        """
+        Calculate embeddings for each sentence group.
+        """
+        embeddings = []
+        for group in sentence_groups:
+            result = self.text_embedder.run(group)
+            embeddings.append(result["embedding"])
+        return embeddings
+
+    def _find_split_points(self, embeddings: List[List[float]]) -> List[int]:
+        """
+        Find split points based on cosine distances between sequential embeddings.
+        """
+        if len(embeddings) <= 1:
+            return []
+
+        # Calculate cosine distances between sequential pairs
+        distances = []
+        for i in range(len(embeddings) - 1):
+            distance = self._cosine_distance(embeddings[i], embeddings[i + 1])
+            distances.append(distance)
+
+        # Calculate threshold based on percentile
+        threshold = np.percentile(distances, self.percentile * 100)
+
+        # Find indices where distance exceeds threshold
+        split_points = []
+        for i, distance in enumerate(distances):
+            if distance > threshold:
+                split_points.append(i + 1)  # +1 because we want to split after this point
+
+        return split_points
+
+    def _cosine_distance(self, embedding1: List[float], embedding2: List[float]) -> float:
+        """
+        Calculate cosine distance between two embeddings.
+        """
+        vec1 = np.array(embedding1)
+        vec2 = np.array(embedding2)
+
+        norm1 = np.linalg.norm(vec1)
+        norm2 = np.linalg.norm(vec2)
```

Comment:
Does this mean we have an implicit numpy dependency? It normally gets installed with things like transformers, torch or sentence transformers but it's not part of the base haystack deps. 

And since some of our embedders use APIs it's not guaranteed that they will come with a numpy dep so we may want to explicitly include it or see if we could avoid using numpy.

Reply:
We do have numpy as a dependency https://github.com/deepset-ai/haystack/blob/2693f39e4430cd83eb6f2860a399b27f6070645a/pyproject.toml#L57

also using it here
https://github.com/deepset-ai/haystack/blob/2693f39e4430cd83eb6f2860a399b27f6070645a/haystack/utils/misc.py#L7

Reply:
Ahh you're right sorry about that. I initially saw it in test deps so I thought that meant it wasn't in main deps. But the testing is only for ensuring we test on version 2 or higher